src/types/mod.rs

```diff
@@ -1,160 +1,219 @@
+pub struct Variable {
+    pub(crate) id: usize,
 }
-
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub enum Atom<V: Variable, O: Operator> {
-    Variable(V),
-    Operator(O),
+impl Variable {
+    pub fn name<'sig>(&self, sig: &'sig Signature) -> Option<&'sig str> {
+        let opt = &sig.variables[self.id];
+        opt.as_ref().map(|s| s.as_str())
+    }
+    pub fn display(&self, sig: &Signature) -> String {
+        if let Some(ref name) = sig.variables[self.id] {
+            name.clone()
+        } else {
+            format!("<var {}>", self.id)
+        }
+    }
 }
 
-/// Represents a symbol with fixed [`Arity`].
+/// A symbol with fixed arity, defined in the signature.
+///
+/// To construct an [`Operator`], use [`Signature::new_op`]
 ///
-/// [`Arity`]: type.Arity_.html
-#[derive(Debug, Clone, Eq)]
-pub struct Op {
-    id: DeBruijn,
-    arity: usize,
-    name: Option<String>,
+/// [`Operator`]: struct.Operator.html
+/// [`Signature::new_op`]: struct.Signature.html#method.new_op
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Operator {
+    pub(crate) id: usize,
 }
-impl Op {
-    pub fn new(id: DeBruijn, arity: usize, name: Option<String>) -> Op {
-        Op { id, arity, name }
+impl Operator {
+    pub fn arity(&self, sig: &Signature) -> u32 {
+        sig.operators[self.id].0
+    }
+    pub fn name<'sig>(&self, sig: &'sig Signature) -> Option<&'sig str> {
+        let opt = &sig.operators[self.id].1;
+        opt.as_ref().map(|s| s.as_str())
     }
-    pub fn name(&self) -> Option<&str> {
-        self.name.as_ref().map(|s| s.as_str())
+    pub fn display(&self, sig: &Signature) -> String {
+        if let (_, Some(ref name)) = sig.operators[self.id] {
+            name.clone()
+        } else {
+            format!("<op {}>", self.id)
+        }
     }
 }
-impl Hash for Op {
-    fn hash<H: Hasher>(&self, state: &mut H) {
-        self.id.hash(state);
-        self.arity.hash(state);
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Atom<'a> {
+    Variable(&'a Variable),
+    Operator(&'a Operator),
+}
+impl<'a> From<&'a Variable> for Atom<'a> {
+    fn from(var: &'a Variable) -> Atom<'a> {
+        Atom::Variable(var)
     }
 }
-impl PartialEq for Op {
-    fn eq(&self, other: &Self) -> bool {
-        self.id == other.id && self.arity == other.arity
+impl<'a> From<&'a Operator> for Atom<'a> {
+    fn from(op: &'a Operator) -> Atom<'a> {
+        Atom::Operator(op)
     }
 }
-impl Operator for Op {
-    fn new_distinct<'a, T>(ops: T, arity: usize, name: Option<String>) -> Op
-    where
-        T: IntoIterator<Item = &'a Op>,
-    {
-        let id = match ops.into_iter().map(|o| o.id).max() {
-            Some(n) => n + 1,
-            _ => 0,
+
+/// Records a universe of symbols.
+#[derive(Clone, Debug)]
+pub struct Signature {
+    /// Stores the (arity, maybe name) for every operator.
+    pub(crate) operators: Vec<(u32, Option<String>)>,
+    /// Stores the name for every variable.
```

Comment:
"maybe name"

Reply:
I was thinking in the haskell sense of `Maybe` 😛

Reply:
Yup - just noting that you say `maybe name` for `operators` but `name` for `variables`. This is a total nitpick, but it would be nice if both said `maybe name`. It's not worth another commit, though!